File: 2024/day14.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <vector>
#include <map>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
"p=0,4 v=3,-3\n"
"p=6,3 v=-1,-3\n"
"p=10,3 v=-1,2\n"
"p=2,0 v=2,-1\n"
"p=0,0 v=1,3\n"
"p=3,0 v=-2,-2\n"
"p=7,6 v=-1,-3\n"
"p=3,0 v=-1,-2\n"
"p=9,3 v=2,3\n"
"p=7,3 v=-1,2\n"
"p=2,4 v=2,-3\n"
"p=9,5 v=-3,-3"
);

bool DEBUG = false;
bool TEST = false;
#define DAY "day14"

int WIDTH=-1;
int HEIGHT=-1;

void printgrid(IntMatrix & data)
{
    uint8_t grid[HEIGHT][WIDTH] = {0};
    bzero( grid, WIDTH*HEIGHT );
    for( auto & bot : data )
        grid[bot[1]][bot[0]] += 1;
    for( int y=0; y < HEIGHT; y++ )
    {   
        for( int x=0; x < WIDTH; x++ )
            cout << ".1234"[grid[y][x]];
        cout << endl;
    }
}

void move(IntMatrix & data)
{
    for( auto & row : data )
    {
        row[0] = (row[0] + row[2] + WIDTH) % WIDTH;
        row[1] = (row[1] + row[3] + HEIGHT) % HEIGHT;
    }
}

int64_t part1(IntMatrix & data)
{
    for( int i = 0; i < 100; i++ )
        move(data);

    int hw = WIDTH/2;
    int hh = HEIGHT/2;
    int k[4] = {0,0,0,0};
    for( auto & row : data )
    {
        if( row[0] < hw && row[1] < hh )
            k[0]++;
        if( row[0] > hw && row[1] < hh )
            k[1]++;
        if( row[0] < hw && row[1] > hh )
            k[2]++;
        if( row[0] > hw && row[1] > hh )
            k[3]++;
    }
    return k[0]*k[1]*k[2]*k[3]; 
}

// The tree is solid, so the standard deviation of the coordinates goes WAY down.
// Tpyical is 30, tree gets 19.

bool detect_tree( IntMatrix & data )
{
    int n = data.size();
    float sumx = 0;
    float sumx2 = 0;
    float sumy = 0;
    float sumy2 = 0;
    for( auto & row : data )
    {
        sumx += row[0];
        sumx2 += row[0]*row[0];
        sumy += row[1];
        sumy2 += row[1]*row[1];
    }
    float meanx = sumx / n;
    float meany = sumy / n;
    float stdx = sqrt((sumx2 / n) - (meanx * meanx));
    float stdy = sqrt((sumy2 / n) - (meany * meany));

    return stdx < 20 && stdy < 20;
}  


int64_t part2(IntMatrix & data)
{
    for( int i = 100; i < 10000; i++ )
    {
        move(data);
        if( detect_tree(data) )
        {
            if( DEBUG )
                printgrid(data);
            return i+1;
        }
    }
    return 0;
}


int main( int argc, char ** argv )
{
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents(DAY".txt");
    if( TEST )
    {
        WIDTH = 11;
        HEIGHT = 7;
    }
    else
    {
        WIDTH = 101;
        HEIGHT = 103;
    }

    // For parsing, each line has four numbers, possibly with a sign.

    IntMatrix data(1);
    int accum = 0;
    int state = 0;
    for( char c : input )
    {
        if( c == '-' )
        {
            state = -1;
        }
        else if( isdigit(c) )
        {
            if( !state )
                state = 1;
            accum = accum * 10 + c - '0';
        }
        else
        {
            if( state )
            {
                data.back().push_back( accum * state );
                accum = 0;
                state = 0;
            }
            if( c == '\n' )
                data.push_back( vector<int>() );
        }
    }

    if( accum )
        data.back().push_back(accum * state);

    cout << "Part 1: " << part1(data) << "\n";
    if( DEBUG )
        printgrid( data );
    if( !TEST )
        cout << "Part 2: " << part2(data) << "\n";
}

File: 2024/day14.go
package main

import (
	"fmt"

	"aoc/tools"
	_ "embed"
	"gonum.org/v1/gonum/stat"
)

var DEBUG bool = false
var TEST bool = false

var test string = `
p=0,4 v=3,-3
p=6,3 v=-1,-3
p=10,3 v=-1,2
p=2,0 v=2,-1
p=0,0 v=1,3
p=3,0 v=-2,-2
p=7,6 v=-1,-3
p=3,0 v=-1,-2
p=9,3 v=2,3
p=7,3 v=-1,2
p=2,4 v=2,-3
p=9,5 v=-3,-3`[1:]

//go:embed day14.txt
var live string

var WIDTH int = -1
var HEIGHT int = -1

func printgrid(data [][]int) {
	grid := make([][]byte, HEIGHT)
	for y := 0; y < HEIGHT; y++ {
		grid[y] = make([]byte, WIDTH)
	}

	for _, bot := range data {
		grid[bot[1]][bot[0]]++
	}

	for _, row := range grid {
		var s string
		for _, c := range row {
			s += string(".1234"[c])
		}
		fmt.Println(s)
	}
}

func move(data [][]int) {
	for _, row := range data {
		row[0] = (row[0] + row[2] + WIDTH) % WIDTH
		row[1] = (row[1] + row[3] + HEIGHT) % HEIGHT
	}
}

// The tree is solid, so the standard deviation of the coordinates goes WAY down.
// Tpyical is 30, tree gets 19.

func detect_tree(data [][]int) bool {
	var x []float64
	var y []float64
	for _, bot := range data {
		x = append(x, float64(bot[0]))
		y = append(y, float64(bot[1]))
	}
	stx := stat.StdDev(x, nil)
	sty := stat.StdDev(y, nil)
	return stx < 20 && sty < 20 
}

func part1(data [][]int) int {
	for range 100 {
		move(data)
	}

	hw := WIDTH / 2
	hh := HEIGHT / 2
	k1, k2, k3, k4 := 0, 0, 0, 0

	for _, bot := range data {
		if bot[0] < hw && bot[1] < hh {
			k1++
		}
		if bot[0] > hw && bot[1] < hh {
			k2++
		}
		if bot[0] < hw && bot[1] > hh {
			k3++
		}
		if bot[0] > hw && bot[1] > hh {
			k4++
		}
	}
	return k1 * k2 * k3 * k4
}

func part2(data [][]int) int {
	var i int
	for i = 100; i < 10000; i++ {
		move(data)
		if detect_tree(data) {
			if DEBUG {
				printgrid(data)
			}
			break
		}
	}
	return i + 1
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := tools.GetNumbers[int](input)

	if TEST {
		WIDTH = 7
		HEIGHT = 11
	} else {
		WIDTH = 101
		HEIGHT = 103
	}

	if DEBUG {
		printgrid(data)
	}

	fmt.Println("Part 1:", part1(data))
	if !TEST {
		fmt.Println("Part 2:", part2(data))
	}
}
